libs/hwes/eventstoredb/projections/custom/custom.go

```diff
@@ -0,0 +1,135 @@
+package custom
+
+import (
+	"context"
+	"fmt"
+	"github.com/EventStore/EventStore-Client-Go/esdb"
+	zlog "github.com/rs/zerolog/log"
+	"hwes"
+	"strings"
+)
+
+const subscriptionGroupName = "custom"
+
+type eventHandler func(ctx context.Context, evt hwes.Event) (error, esdb.Nack_Action)
+
+type CustomProjection struct {
+	es            *esdb.Client
+	eventHandlers map[string]eventHandler
+}
+
+func NewCustomProjection(esdbClient *esdb.Client) *CustomProjection {
+	return &CustomProjection{es: esdbClient, eventHandlers: make(map[string]eventHandler, 0)}
+}
+
+func (p *CustomProjection) RegisterEventListener(eventType string, eventHandler eventHandler) *CustomProjection {
+	if _, found := p.eventHandlers[eventType]; found {
+		zlog.Error().
+			Str("eventType", eventType).
+			Msg("override existing event handler")
+	}
+
+	p.eventHandlers[eventType] = eventHandler
+	return p
+}
+
+func (p *CustomProjection) handleEvent(ctx context.Context, event hwes.Event) (error, esdb.Nack_Action) {
+	eventHandler, found := p.eventHandlers[event.EventType]
+	if !found {
+		return fmt.Errorf("event type '%s' is invalid", event.EventType), esdb.Nack_Unknown
+	}
+	return eventHandler(ctx, event)
+}
+
+func (p *CustomProjection) Subscribe(ctx context.Context) error {
+	log := zlog.Ctx(ctx)
+
+	// Create subscription on EventStoreDB
+	persistentAllSubscriptionOptions := esdb.PersistentAllSubscriptionOptions{}
+	err := p.es.CreatePersistentSubscriptionAll(ctx, subscriptionGroupName, persistentAllSubscriptionOptions)
+	if err != nil {
+		// Ignore subscription already exists error
```

Comment:
why is this ok to do and why are we doing it? (add a comment)

will this cause race-conditions?

Reply:
No race-conditions. If the initial creation fails, we fail on connection as well.

Reply:
What will happen when two services are subscribing after another?
Both at connect-time and at event-handle-time

Won't both try to handle the same event at the same time?

Reply:
Thats why we are using [Persistent subscriptions](https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html) as well. Every projection gets grouped by its `subscriptionGroupName` (after my refactor from today). Every group has a at-least-once guarantee.